objects/src/accounts/storage.rs

```diff
@@ -1,47 +1,83 @@
+/// will allow the user to store any Merkle structures they need.  This is achieved by storing the
+/// root of the Merkle structure as a leaf in the simple sparse merkle tree. When `AccountStorage`
+/// is serialized it will check to see if any of the leafs in the simple sparse Merkle tree are
+/// Merkle roots of other Merkle structures.  If any Merkle roots are found then the Merkle
+/// structures will be persisted in the `AccountStorage` `MerkleStore`.
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct AccountStorage {
-    // TODO: add compact SMT as backing storage
+    slots: SimpleSmt,
+    store: MerkleStore,
 }
 
 impl AccountStorage {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// Depth of the storage tree.
+    pub const STORAGE_TREE_DEPTH: u8 = 8;
+
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
     /// Returns a new instance of account storage initialized with the provided items.
-    pub fn new(_items: &[StorageItem]) -> Self {
-        Self {}
+    pub fn new(
+        items: Vec<StorageItem>,
+        store: MerkleStore,
+    ) -> Result<AccountStorage, AccountError> {
+        let slots = SimpleSmt::new(Self::STORAGE_TREE_DEPTH)
+            .expect("depth is valid")
+            .with_leaves(items.into_iter().map(|x| (x.0 as u64, x.1)))
+            .expect("`StorageItem` index is u8 - index within range");
+
+        Ok(Self { slots, store })
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
     /// Returns a commitment to this storage.
-    pub fn root(&self) -> Digest {
-        Digest::default()
+    pub fn root(&self) -> Word {
+        self.slots.root()
     }
 
-    /// Returns an item from the storage at the specified key. Both items and keys are Words which
-    /// consist of 4 field elements.
+    /// Returns an item from the storage at the specified index.
     ///
     /// If the item is not present in the storage, [ZERO; 4] is returned.
-    pub fn get_item(&self, key: Word) -> Word {
-        // the index of the item is the hash of its key
-        let _index: Word = Hasher::merge(&[key.into(), [ZERO; 4].into()]).into();
-        todo!("retrieve the item from the tree");
+    pub fn get_item(&self, index: u8) -> Result<Word, AccountError> {
+        let item_index =
+            NodeIndex::new(Self::STORAGE_TREE_DEPTH, index as u64).expect("u8 is valid");
+        self.slots.get_node(item_index).map_err(AccountError::GetStorageItemError)
+    }
+
+    /// Sets an item from the storage at the specified index.
+    pub fn set_item(&mut self, index: u8, value: Word) -> Result<(), AccountError> {
+        self.slots
+            .insert_leaf(index as u64, value)
+            .map_err(AccountError::SetStorageItemError)
+    }
```

Comment:
Similar comment to the previous one: maybe this method should be infallible?

Reply:
Why doesn't `insert_leaf` return the previous leaf value for the `SimpleSmt`?  At the minute we have an inconsistency between masm and rust account storage API's.  The masm `set_item` procedure returns the old value, the rust `set_item` function does not.

Reply:
No specific reason - I think it could return old leaf just as well (assuming it doesn't complicate things too much).

Reply:
We can do this quite simply in `AccountStorage::set_item(..)` by calling `get_item(..)` as follows:

```rust
    /// Sets an item from the storage at the specified index.
    pub fn set_item(&mut self, index: u8, value: Word) -> Word {
        let old_value = self.get_item(index);
        self.slots
            .insert_leaf(index as u64, value)
            .expect("index is u8 - index within range");
        old_value
    }
```

I think we can do this for now.  My expectation would be that the `SimpleSmt` would return the old value upon execution of `insert_leaf(..)`.  I will create an issue in `crypto` to discuss this.

Reply:
updated